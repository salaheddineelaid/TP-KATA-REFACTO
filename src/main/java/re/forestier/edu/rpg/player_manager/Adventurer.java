package re.forestier.edu.rpg.player_manager;

import re.forestier.edu.rpg.itm_manager.Item;

import java.util.ArrayList;
import java.util.HashMap;

public class Adventurer extends Player {
    public Adventurer(String playerName, String avatarName, int money, ArrayList<Item> inventory) {
        super(playerName, avatarName, "ADVENTURER", money, inventory);
    }

    @Override
    protected HashMap<String, Integer> initializeAbilities() {
        HashMap<String, Integer> abilities = new HashMap<>();
        abilities.put("ATK", 3);
        abilities.put("DEF", 1);
        abilities.put("INT", 1); // Intelligence for Adventurer
        return abilities;
    }

    @Override
    protected HashMap<String, Integer> initializeAbilitiesForLevel(int level) {
        HashMap<String, Integer> abilities = new HashMap<>();
        switch (level) {
            case 1:
                abilities.put("ATK", 1);
                abilities.put("DEF", 1);
                abilities.put("INT", 1);
                break;
            case 2:
                abilities.put("ATK", 2);
                abilities.put("DEF", 1);
                abilities.put("INT", 2);
                break;
            case 3:
                abilities.put("ATK", 3);
                abilities.put("DEF", 2);
                abilities.put("INT", 2);
                break;
            case 4:
                abilities.put("ATK", 4);
                abilities.put("DEF", 2);
                abilities.put("INT", 3);
                break;
            case 5:
                abilities.put("ATK", 5);
                abilities.put("DEF", 3);
                abilities.put("INT", 4);
                break;
            default:
                break;
        }
        return abilities;
    }
    @Override
    public void recoverHealth() {
        if (currenthealthpoints < healthpoints / 2) {
            currenthealthpoints += 2; // Adventurer's recovery
            if (retrieveLevel() < 3) {
                currenthealthpoints -= 1; // Penalty for low level
            }
        }
    }

    // Adventurer-specific methods
}