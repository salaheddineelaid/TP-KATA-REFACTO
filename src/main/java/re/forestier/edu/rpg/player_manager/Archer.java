package re.forestier.edu.rpg.player_manager;

import re.forestier.edu.rpg.itm_manager.Item;

import java.util.ArrayList;
import java.util.HashMap;

public class Archer extends Player {
    public Archer(String playerName, String avatarName, int money, ArrayList<Item> inventory) {
        super(playerName, avatarName, "ARCHER", money, inventory);
    }

    @Override
    protected HashMap<String, Integer> initializeAbilities() {
        HashMap<String, Integer> abilities = new HashMap<>();
        abilities.put("ATK", 3);
        abilities.put("DEF", 1);
        abilities.put("AGI", 2); // Agility for Archer
        return abilities;
    }

    @Override
    protected HashMap<String, Integer> initializeAbilitiesForLevel(int level) {
        HashMap<String, Integer> abilities = new HashMap<>();
        switch (level) {
            case 1:
                abilities.put("ATK", 1);
                abilities.put("AGI", 1);
                break;
            case 2:
                abilities.put("ATK", 2);
                abilities.put("AGI", 1);
                break;
            case 3:
                abilities.put("ATK", 3);
                abilities.put("AGI", 2);
                break;
            case 4:
                abilities.put("ATK", 3);
                abilities.put("AGI", 3);
                break;
            case 5:
                abilities.put("ATK", 4);
                abilities.put("AGI", 4);
                break;
            default:
                break;
        }
        return abilities;
    }

     @Override
    protected void recoverHealth() {
        if (currenthealthpoints < healthpoints / 2) {
            currenthealthpoints += 1; // Basic recovery

            // Additional recovery if the Archer has a Magic Bow
            if (inventory.contains("Magic Bow")) {
                currenthealthpoints += currenthealthpoints / 8 - 1;
            }
        }
    }

    // Archer-specific methods
}