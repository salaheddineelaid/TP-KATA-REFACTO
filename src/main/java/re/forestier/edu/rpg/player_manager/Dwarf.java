package re.forestier.edu.rpg;

import java.util.ArrayList;
import java.util.HashMap;
public class Dwarf extends Player {
    public Dwarf(String playerName, String avatarName, int money, ArrayList<String> inventory) {
        super(playerName, avatarName, "DWARF", money, inventory);
    }

    @Override
    protected HashMap<String, Integer> initializeAbilities() {
        HashMap<String, Integer> abilities = new HashMap<>();
        abilities.put("ATK", 4);
        abilities.put("DEF", 2);
        abilities.put("ALC", 3); // Alchemy for Dwarf
        return abilities;
    }

    @Override
    protected HashMap<String, Integer> initializeAbilitiesForLevel(int level) {
        HashMap<String, Integer> abilities = new HashMap<>();
        switch (level) {
            case 1:
                abilities.put("ATK", 1);
                abilities.put("DEF", 1);
                abilities.put("ALC", 1);
                break;
            case 2:
                abilities.put("ATK", 2);
                abilities.put("DEF", 2);
                abilities.put("ALC", 2);
                break;
            case 3:
                abilities.put("ATK", 3);
                abilities.put("DEF", 2);
                abilities.put("ALC", 3);
                break;
            case 4:
                abilities.put("ATK", 4);
                abilities.put("DEF", 3);
                abilities.put("ALC", 4);
                break;
            case 5:
                abilities.put("ATK", 5);
                abilities.put("DEF", 4);
                abilities.put("ALC", 5);
                break;
            default:
                break;
        }
        return abilities;
    }
    @Override
    protected void recoverHealth() {
        if (currentHealthPoints < healthPoints / 2) {
            if (inventory.contains("Holy Elixir")) {
                currentHealthPoints += 1; // Recover with Holy Elixir
            }
            currentHealthPoints += 1; // Basic recovery for Dwarf
        }
    }

    // Dwarf-specific methods
}
