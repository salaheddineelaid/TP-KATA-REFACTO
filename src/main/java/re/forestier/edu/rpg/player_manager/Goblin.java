package re.forestier.edu.rpg.player_manager;

import re.forestier.edu.rpg.itm_manager.Item;

import java.util.ArrayList;
import java.util.HashMap;

public class Goblin extends Player {
    public Goblin(String playerName, String avatarName, int money, ArrayList<Item> inventory) {
        super(playerName, avatarName, "GOBLIN", money, inventory);
    }

    @Override
    protected HashMap<String, Integer> initializeAbilities() {
        HashMap<String, Integer> abilities = new HashMap<>();
        abilities.put("ATK", 2);
        abilities.put("DEF", 0);
        abilities.put("INT", 2);
        abilities.put("ALC", 1);
        return abilities;
    }

    @Override
    protected HashMap<String, Integer> initializeAbilitiesForLevel(int level) {
        HashMap<String, Integer> abilities = new HashMap<>();
        switch (level) {
            case 1:
                abilities.put("ATK", 2);
                abilities.put("INT", 2);
                abilities.put("ALC", 1);
                break;
            case 2:
                abilities.put("ATK", 3);
                abilities.put("ALC", 4);
                break;
            case 3:
                abilities.put("VIS", 1);
                break;
            case 4:
                abilities.put("DEF", 1);
                break;
            case 5:
                abilities.put("DEF", 2);
                abilities.put("ATK", 4);
                break;
            default:
                break;
        }
        return abilities;
    }

    @Override
    protected void recoverHealth() {
        if (currenthealthpoints < healthpoints / 2) {
            currenthealthpoints += 1;
        }
    }


}