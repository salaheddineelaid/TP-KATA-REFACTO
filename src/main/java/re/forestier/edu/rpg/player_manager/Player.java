package re.forestier.edu.rpg.player_manager;

import re.forestier.edu.rpg.itm_manager.Item;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Random;

public abstract  class Player {
    public String playerName;
    public String avatarName;
    private String avatarClass;

    public Integer money;
    private Float __real_money__;


    public int level;
    public int healthpoints;
    public int currenthealthpoints;
    public int xp;
    protected double maxWeight; // Poids maximal que le joueur peut porter
    protected double currentWeight; // Poids actuel dans l'inventaire


    public HashMap<String, Integer> abilities;
    public ArrayList<Item> inventory;
     public Player(String playerName, String avatarName, String avatarClass, int money, double maxWeight, ArrayList<Item> inventory) {
        this.playerName = playerName;
        this.avatarName = avatarName;
        this.avatarClass = avatarClass;
        this.money = money;
        this.inventory = inventory;
        this.level = 1; // Start at level 1
        this.xp = 0; // Start with 0 XP
        this.abilities = initializeAbilities();
        this.currenthealthpoints = 100;
        this.healthpoints=100;
        this.maxWeight = maxWeight;

         this.currentWeight = 0;
    }

    public Player(String playerName, String avatarName, String avatarClass, int money, ArrayList<Item> inventory) {
        this.playerName = playerName;
        this.avatarName = avatarName;
        this.avatarClass = avatarClass;
        this.money = money;
        this.inventory = inventory;
        this.level = 1; // Start at level 1
        this.xp = 0; // Start with 0 XP
        this.abilities = initializeAbilities();
        this.currenthealthpoints = 100;
        this.healthpoints=100;
        this.maxWeight = 50;

        this.currentWeight = 0;
    }


    public void removeMoney(int amount) throws IllegalArgumentException {
        if (money - amount < 0) {
            throw new IllegalArgumentException("Player can't have a negative money!");
        }

        money = Integer.parseInt(money.toString()) - amount;
    }

    public void addItem(Item item) throws IllegalArgumentException {
        if (currentWeight + item.getWeight() > maxWeight) {
            throw new IllegalArgumentException("Cannot add item: weight limit exceeded!");
        }
        inventory.add(item);
        currentWeight += item.getWeight();
    }
    public void addMoney(int amount) {
        var value = Integer.valueOf(amount);
        money = money + (value != null ? value : 0);
    }

    protected abstract HashMap<String, Integer> initializeAbilities();

    public void addXp(int xp) {
        this.xp += xp;
        // Level up logic
        int newLevel = retrieveLevel();
        if (newLevel > level) {
            level = newLevel;
            gainAbility();
            gainItem();
        }
    }

    private void gainItem() {
        Item[] objectList = {
                new Item("Lookout Ring", "A magical ring that enhances vision.", 0.5, 100),
                new Item("Scroll of Stupidity", "A scroll that confuses enemies.", 0.2, 50),
                new Item("Draupnir", "A magical ring that multiplies wealth.", 0.3, 200),
                new Item("Magic Charm", "A charm that grants luck.", 0.1, 75),
                new Item("Rune Staff of Curse", "A staff that curses enemies.", 1.0, 150),
                new Item("Combat Edge", "A blade that sharpens fighting skills.", 0.8, 125),
                new Item("Holy Elixir", "An elixir that restores health.", 0.4, 200)
        };
        Random random = new Random();
        addItem(objectList[random.nextInt(objectList.length)]);
    }


    protected void gainAbility() {
        HashMap<String, Integer> newAbilities = initializeAbilitiesForLevel(level);
        abilities.forEach((ability, value) -> abilities.put(ability, abilities.get(ability) + newAbilities.getOrDefault(ability, 0)));
    }

    protected abstract HashMap<String, Integer> initializeAbilitiesForLevel(int level);

    public int retrieveLevel() {
        // Level calculation logic based on xp
        if (xp < 10) return 1;
        else if (xp < 27) return 2;
        else if (xp < 57) return 3;
        else if (xp < 111) return 4;
        return 5;
    }
    public int getXp() {
        return this.xp;
    }

    public void majFinDeTour() {
        if (currenthealthpoints <= 0) {
            System.out.println("Le joueur est KO !");
            return;
        }

        // Call the subclass-specific recovery method
        recoverHealth();

        // Cap current health points to maximum health points
        if (currenthealthpoints > healthpoints) {
            currenthealthpoints = healthpoints;
        }
    }

    protected abstract void recoverHealth();

    

}