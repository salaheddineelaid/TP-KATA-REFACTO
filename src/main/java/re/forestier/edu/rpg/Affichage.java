package re.forestier.edu.rpg;

import re.forestier.edu.rpg.player_manager.Player;

public class Affichage {

    public static String afficherJoueur(Player player) {
        final String[] finalString = {"Joueur " + player.avatarName + " joué par " + player.playerName};
        finalString[0] += "\nNiveau : " + player.retrieveLevel() + " (XP totale : " + player.xp + ")";
        finalString[0] += "\n\nCapacités :";
        player.abilities.forEach((name, level) -> {
            finalString[0] += "\n   " + name + " : " + level;
        });
        finalString[0] += "\n\nInventaire :";
        player.inventory.forEach(item -> {
            finalString[0] += "\n   " + item.getName();
        });

        return finalString[0];
    }

    public static String afficherJoueurMarkdown(Player player) {
        StringBuilder markdown = new StringBuilder();
        markdown.append("# Joueur ").append(player.avatarName).append(" joué par ").append(player.playerName).append("\n");
        markdown.append("## Niveau : ").append(player.retrieveLevel()).append(" (XP totale : ").append(player.xp).append(")\n\n");

        markdown.append("**Capacités :**\n");
        player.abilities.forEach((name, level) -> {
            markdown.append("* ").append(name).append(" : ").append(level).append("\n");
        });

        markdown.append("\n**Inventaire :**\n");
        player.inventory.forEach(item -> {
            markdown.append("* ").append(item.getName()).append("\n");
        });

        return markdown.toString();
    }

}
