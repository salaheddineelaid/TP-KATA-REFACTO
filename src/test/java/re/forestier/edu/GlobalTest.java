package re.forestier.edu;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import re.forestier.edu.rpg.Affichage;
import re.forestier.edu.rpg.player_manager.Adventurer;


import java.util.ArrayList;

import static org.approvaltests.Approvals.verify;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;
import static org.junit.jupiter.api.Assertions.fail;

public class GlobalTest {

    @Test
    void testAffichageBase() {
        Adventurer player = new Adventurer("Florian", "Gnognak le Barbare",  200, new ArrayList<>());
        player.addXp(20);
        player.inventory = new ArrayList<>();

        verify(Affichage.afficherJoueur(player));
    }
}
