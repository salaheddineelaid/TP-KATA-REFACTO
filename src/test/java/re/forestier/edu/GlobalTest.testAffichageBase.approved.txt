Joueur Gnognak le Barbare joué par Florian
Niveau : 2 (XP totale : 20)

Capacités :
   DEF : 2
   ATK : 5
   INT : 3

Inventaire :